### PR TITLE
Optimizations for large builds

### PR DESCRIPTION
**Pull request by aras-p**
Redo how the data is gathered & stored to better scale for large builds (ref #10, #30).

TLDR: **3x faster, 15x lower memory usage, 4x smaller file size**.

Previously, `--stop` was just smashing all input JSON files together, and `--analyze` was parsing that huge JSON file, and performing the analysis. The memory allocator used was a simple "bump a pointer, never free" allocator. On a decent size project (e.g. Unity editor build), that was taking 4.8sec / 5.8GB for stop, and 9.9sec / 7.3GB for analyze (so **total 14.7sec**, and max **7.3GB memory** usage). The resulting file was **1.01GB size**.

Now:

- [x] do the JSON parsing during `--stop` step into the build events data structures (which includes string/name deduplication), and as a result just store the binary data structure file. `--analyze` reads that binary file and does no JSON parsing at all now.
- [x] use regular built-in memory allocator (saves a ton of memory for large analyses).
- [x] multi-threaded `--stop` part that parses all the JSON files, using [enkiTS](https://github.com/dougbinks/enkiTS) task scheduler.
- [x] replace `sajson` JSON parser with [simdjson](https://github.com/lemire/simdjson).
- [x] use [bytell_hash_map](https://probablydance.com/2018/05/28/a-new-fast-hash-table-in-response-to-googles-new-fast-hash-table/) instead of C++ built-in one.
- [x] use [xxHash](https://github.com/Cyan4973/xxHash) hash function instead of C++ built-in one.
- [x] optimize "Template sets" and "Function sets" analysis calculation

On the same Unity editor build: 2.4sec / 0.35GB for stop, and 2.7sec / 0.51GB for analyze (so **total 5.1sec**, and max **0.5GB memory** usage). The resulting file is **280MB size**.

However the data files produced by `--stop` and used by `--analyze` are no longer compatible with previous versions (previous were just JSON, now custom binary).

### CODE REVIEW

File: src/Analysis.cpp [hunk 1]
```diff
@@ -360,9 +361,17 @@ void Analysis::EmitCollapsedTemplates()
 void Analysis::EmitCollapsedTemplateOpt()
 {
     std::unordered_map<std::string, InstantiateEntry> collapsed;
+    std::unordered_map<DetailIndex, std::string> collapsedNameCache;
+    
     for (const auto& fn : functions)
     {
-        auto &stats = collapsed[collapseName(llvm::demangle(GetBuildName(fn.first.first)))];
```
Comment (ben-craig) [inline src/Analysis.cpp:365]: I guess this was showing up in profiles?

When writing this, it did cross my mind that I was doing potentially redundant demangling.  I didn't cache the demangling at the time though, because I was worried about the memory consumption.  If that looks fine on Unity, then it looks fine to me.

File: src/Analysis.cpp [hunk 2]
```diff
@@ -449,7 +458,7 @@ void Analysis::EndAnalysis()
             if (dname.size() > config.maxName)
                 dname = dname.substr(0, config.maxName-2) + "...";
             int ms = int(e.second.us / 1000);
-            int avg = int(ms / e.second.count);
+            int avg = int(ms / std::max(e.second.count,1));
```
Comment (ben-craig) [inline src/Analysis.cpp:441]: I guess I just never ran into the div zero case... good catch.

**Review by ben-craig — COMMENTED** [review]
aras-p doesn't optimize runtime code.
aras-p doesn't optimize build throughput.
aras-p optimizes the build throughput profiler.

I'm now expecting performance patches for whatever perf visualizer you use.

File: src/Analysis.cpp (hunk 1 above)
Comment (aras-p) [reply src/Analysis.cpp:365]: It was showing up a bit indeed (not much). I've checked that the produced results are the same in all tests & in the full unity build analysis, so Should Be Fine :)
